File: frontend/src/types/category.ts
/**
 * 카테고리 타입 정의 (v2.0 - 소유자 기반 카테고리)
 */

/**
 * 카테고리 기본 정보
 */
export interface Category {
  categoryId: number
  categoryCode: string
  categoryName: string
  description?: string
  categoryColor?: string
  sortOrder: number
  useYn: string
  ownerUsername: string
  ownerName?: string
  createdAt?: string
  createdBy?: string
  updatedAt?: string
  updatedBy?: string
}

/**
 * 카테고리 상세 정보 (속성 및 공유 포함)
 */
export interface CategoryDetail extends Category {
  properties: CategoryProperty[]
  shares: CategoryShare[]
  boardCount?: number
}

/**
 * 카테고리에 연결된 속성 (API 응답 형식)
 */
export interface CategoryProperty {
  categoryPropertyId: number
  categoryId: number
  propertyId: number
  propertyName: string
  propertyType: string
  ownerType?: string  // GLOBAL, MANAGER, USER
  requiredYn?: string
  sortOrder: number
  defaultValue?: string
  createdAt?: string
  createdBy?: string
}

/**
 * 카테고리 공유 정보 (API 응답 형식)
 */
export interface CategoryShare {
  shareId: number
  categoryId?: number
  shareType: 'USER' | 'DEPARTMENT'
  shareTarget: string  // USERNAME 또는 DEPARTMENT_CODE
  shareTargetName?: string  // 사용자명 또는 부서명
  permission: 'READ' | 'EDIT'
  sharedAt?: string
  sharedBy?: string
}

/**
 * 카테고리 생성 요청
 */
export interface CategoryCreateRequest {
  categoryCode: string
  categoryName: string
  description?: string
  categoryColor?: string
  sortOrder?: number
  propertyIds?: number[]
}

/**
 * 카테고리 수정 요청
 */
export interface CategoryUpdateRequest {
  categoryCode?: string
  categoryName?: string
  description?: string
  categoryColor?: string
  sortOrder?: number
  useYn?: string
}

/**
 * 카테고리 공유 추가 요청
 */
export interface CategoryShareRequest {
  shareType: 'USER' | 'DEPARTMENT'
  shareTarget: string  // USERNAME 또는 DEPARTMENT_CODE
  permission: 'READ' | 'EDIT'
}

/**
 * 카테고리 속성 추가 요청
 */
export interface CategoryPropertyRequest {
  propertyId: number
  sortOrder?: number
  defaultValue?: string
}

/**
 * 카테고리 속성 순서 변경 요청
 */
export interface CategoryPropertyOrderRequest {
  propertyId: number
  sortOrder: number
}

/**
 * 카테고리 속성 기본값 변경 요청
 */
export interface CategoryPropertyDefaultValueRequest {
  propertyId: number
  defaultValue: string
}

File: backend/src/main/java/com/taskflow/domain/Category.java
package com.taskflow.domain;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.List;

/**
 * 카테고리 (속성 그룹) 엔티티
 *
 * 테이블: TB_CATEGORY
 *
 * v2.0 변경사항:
 * - categoryCode, description, ownerUsername 필드 추가
 * - 사용자별 카테고리 소유 구조
 * - 카테고리 공유 기능 지원
 */
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Category {

    /**
     * 카테고리 ID (PK)
     */
    private Long categoryId;

    /**
     * 카테고리 코드
     */
    private String categoryCode;

    /**
     * 카테고리명
     */
    private String categoryName;

    /**
     * 설명
     */
    private String description;

    /**
     * 표시 색상 (#RRGGBB)
     */
    private String categoryColor;

    /**
     * 소유자 USERNAME
     */
    private String ownerUsername;

    /**
     * 정렬 순서
     */
    private Integer sortOrder;

    /**
     * 사용 여부 (Y/N)
     */
    private String useYn;

    /**
     * 생성일시
     */
    private LocalDateTime createdAt;

    /**
     * 생성자 USERNAME
     */
    private String createdBy;

    /**
     * 수정일시
     */
    private LocalDateTime updatedAt;

    /**
     * 수정자 USERNAME
     */
    private String updatedBy;

    /**
     * 삭제일시 (논리삭제)
     */
    private LocalDateTime deletedAt;

    /**
     * 삭제자 USERNAME
     */
    private String deletedBy;

    // ============================================
    // 연관 데이터 (조회용)
    // ============================================

    /**
     * 카테고리에 배정된 속성 목록
     */
    private List<CategoryProperty> properties;

    /**
     * 카테고리 공유 목록
     */
    private List<CategoryShare> shares;

    /**
     * 소유자 이름 (조회용)
     */
    private String ownerName;

    /**
     * 속성 개수 (조회용)
     */
    private Integer propertyCount;

    // =============================================
    // 하위 호환성을 위한 getter (color -> categoryColor)
    // =============================================

    /**
     * 색상 getter (하위 호환성)
     */
    public String getColor() {
        return categoryColor;
    }

    /**
     * 색상 setter (하위 호환성)
     */
    public void setColor(String color) {
        this.categoryColor = color;
    }

    // =============================================
    // 편의 메서드
    // =============================================

    /**
     * 활성 카테고리 여부
     */
    public boolean isActive() {
        return "Y".equals(useYn);
    }
}
